File: 04-vector-search/Solution/python/SearchComponents/indexes.py
# Function to create vector indexes in a MongoDB collection
def createVectorIndexes(collection, index_list, db, collection_name):
    # Get information about the existing indexes in the collection
    collection_indexes = collection.index_information()    

    # Iterate over each index in the index_list
    for indexname, vectorColumn in index_list:
        # Iterate over each index in the collection indexes
        for index in collection_indexes:
            # If the index already exists in the collection
            if index == indexname:
                # Drop the existing index
                collection.drop_index(indexname)
                break
        
            # Create a new IVF index in the collection
            # The index is created using the MongoDB command function
            # The command specifies the collection to create the index in, the name of the index, 
            # the key to index on, and the options for the CosmosDB search        
            db.command({
            'createIndexes': collection_name,
            'indexes': [
                {
                'name': indexname,
                'key': {
                    f"{vectorColumn}": "cosmosSearch"
                },
                'cosmosSearchOptions': {
                    'kind': 'vector-ivf',
                    'numLists': 1,
                    'similarity': 'COS',
                    'dimensions': 1536
                }
                }
            ]
            })

File: 05-deploy-with-aks/node.js/SearchComponents/indexes.js
// Asynchronous function to create vector indexes in a MongoDB collection
async function createVectorIndexes(collection, indexList, db, collectionName) {

    // Get the current indexes in the collection
    const collectionIndexes = await collection.indexInformation();

    // Iterate over each index in the indexList
    for (let [indexName, vectorColumn] of indexList) {
        // Iterate over each index in the collection
        for (let index of Object.keys(collectionIndexes)) {
            // If the index already exists in the collection, drop it
            if (index === indexName) {
                await collection.dropIndex(indexName);
                break;
            }
        }

        // Create a new IVF index in the collection
        // The index is created using the MongoDB command function
        // The command specifies the collection to create the index in, the name of the index, 
        // the key to index on, and the options for the CosmosDB search
        const commandResult = await db.command({
            'createIndexes': collectionName,
            'indexes': [
                {
                    'name': indexName,
                    'key': {
                        [vectorColumn]: "cosmosSearch"
                    },
                    'cosmosSearchOptions': {
                        'kind': 'vector-ivf',
                        'numLists': 1,
                        'similarity': 'COS',
                        'dimensions': 1536
                    }
                }
            ]
        });
    } 

}

// Export the createVectorIndexes function
module.exports.createVectorIndexes = createVectorIndexes;